### PR TITLE
drivers: display: GC9A01A: Add driver for GC9A01A

### PR DESCRIPTION
**Pull request by amrithvenkat**
Adding the driver for the GC9A01A 240x240 1.28 inch round LCD display.

### CODE REVIEW

File: drivers/display/display_gc9a01a.c [hunk 1]
```diff
@@ -0,0 +1,599 @@
+/**
+ * Copyright (c) 2017 Jan Van Winkel <jan.van_winkel@dxplore.eu>
```
Comment (faxe1008) [inline drivers/display/display_gc9a01a.c:2]: Don't copy other peoples copyright. Happened today to me too 
:^)

File: drivers/display/display_gc9a01a.c [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+int gc9a01a_transmit(const struct device *dev, uint8_t cmd, const void *tx_data, size_t tx_len)
+{
+	const struct gc9a01a_config *config = dev->config;
+
+	int r;
+	struct spi_buf tx_buf;
+	struct spi_buf_set tx_bufs = {.buffers = &tx_buf, .count = 1U};
+
+	/* send command */
+	tx_buf.buf = &cmd;
```
Comment (faxe1008) [inline drivers/display/display_gc9a01a.c:76]: Why not set those in the declaration also as you've done above for `tx_bufs`?

File: drivers/display/display_gc9a01a.c [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+	tx_buf.len = 1U;
+
+	gpio_pin_set_dt(&config->cmd_data, GC9A01A_CMD);
```
Comment (faxe1008) [inline drivers/display/display_gc9a01a.c:79]: Maybe also checking the return value here?

File: drivers/display/display_gc9a01a.c [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+		return -ENOTSUP;
+	}
+
+	r = gc9a01a_transmit(dev, GC9A01A_PIXFMT, &tx_data, 1U);
+	if (r < 0) {
+		return r;
+	}
+
+	data->pixel_format = pixel_format;
+	data->bytes_per_pixel = bytes_per_pixel;
```
Comment (faxe1008) [inline drivers/display/display_gc9a01a.c:200]: You can use `BYTES_PER_PIXEL` here directly.

File: drivers/display/display_gc9a01a.c [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+	return 0;
+}
+
+/**
+ * @brief To do overall display device configuration.
+ *
+ * @param dev GC9A01A device instance.
+ * @return 0 when successful, errno otherwise.
+ */
+static int gc9a01a_configure(const struct device *dev)
```
Comment (faxe1008) [inline drivers/display/display_gc9a01a.c:246]: Maybe call it `_init` instead?

File: drivers/display/display_gc9a01a.c [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+	enum display_orientation orientation;
+
+	/* pixel format */
+	if (config->pixel_format == GC9A01A_PIXEL_FORMAT_RGB565) {
+		pixel_format = PIXEL_FORMAT_RGB_565;
+	} else {
+		pixel_format = PIXEL_FORMAT_RGB_888;
+	}
+
+	r = gc9a01a_set_pixel_format(dev, pixel_format); // Set pixel format.
```
Comment (faxe1008) [inline drivers/display/display_gc9a01a.c:261]: Why not pass the cfg pixel format directly here and prevent bad values via `BUILD_CHECK`?

File: drivers/display/display_gc9a01a.c [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+
+	/* orientation */
+	if (config->rotation == 0U) {
+		orientation = DISPLAY_ORIENTATION_NORMAL;
+	} else if (config->rotation == 90U) {
+		orientation = DISPLAY_ORIENTATION_ROTATED_90;
+	} else if (config->rotation == 180U) {
+		orientation = DISPLAY_ORIENTATION_ROTATED_180;
+	} else {
+		orientation = DISPLAY_ORIENTATION_ROTATED_270;
```
Comment (faxe1008) [inline drivers/display/display_gc9a01a.c:274]: Same here. Also I guess a DT enum can be used to prevent bad values here, and if not again `BuILD_CHECK`.

**Review by faxe1008 — CHANGES_REQUESTED** [review]

File: drivers/display/display_gc9a01a.c (hunk 1 above)
Comment (amrithvenkat) [reply drivers/display/display_gc9a01a.c:2]: Thanks for the heads up. will remove with the next commit.

File: drivers/display/display_gc9a01a.c [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+	const struct gc9a01a_config *config = dev->config;
+
+	int r;
+
+	if (!spi_is_ready_dt(&config->spi)) {
+		LOG_ERR("SPI device is not ready");
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(config->cmd_data.port)) {
```
[the remark this answers is not in the document]
Comment (amrithvenkat) [reply drivers/display/display_gc9a01a.c:334]: will change in the next commit

File: drivers/display/display_gc9a01a.c (hunk 5 above)
Comment (amrithvenkat) [reply drivers/display/display_gc9a01a.c:246]: I followed the same pattern as the ilitek drivers. There is already a function `gc9a01a_init()` and the` _configure` is called within the `_init`

File: drivers/display/display_gc9a01a.c (hunk 7 above)
Comment (amrithvenkat) [reply drivers/display/display_gc9a01a.c:274]: Could you point me to a resource or example. I am lost here by what you mean

File: drivers/display/display_gc9a01a.c (hunk 2 above)
Comment (amrithvenkat) [reply drivers/display/display_gc9a01a.c]: changed

File: drivers/display/display_gc9a01a.c (hunk 3 above)
Comment (amrithvenkat) [reply drivers/display/display_gc9a01a.c]: added

File: drivers/display/display_gc9a01a.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (amrithvenkat) [reply drivers/display/display_gc9a01a.c]: done

File: drivers/display/Kconfig.gc9a01a [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# GC9A01A display driver configuration options
+
+# SPDX-License-Identifier: Apache-2.0
+
+config GC9A01A
+	bool "GC9A01A display driver"
+	default y
+	depends on DT_HAS_WAVESHARE_GC9A01A_ENABLED
```
Comment (jfischer-no) [inline drivers/display/Kconfig.gc9a01a:8]: We do not implement drivers for displays or display resellers, but for display controller. Here, it cannot be `DT_HAS_WAVESHARE_GC9A01A_ENABLED`, maybe `DT_HAS_GALAXYCORE_GC9A01A_ENABLED`

File: drivers/display/display_gc9a01a.h [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+#define GC9A01A_PASET 0x2B ///< Page Address Set
+#define GC9A01A_RAMWR 0x2C ///< Memory Write
+
+#define GC9A01A_PTLAR 0x30    ///< Partial Area
+#define GC9A01A_VSCRDEF 0x33  ///< Vertical Scrolling Definition
+#define GC9A01A_TEOFF 0x34    ///< Tearing effect line off
+#define GC9A01A_TEON 0x35     ///< Tearing effect line on
+#define GC9A01A_MADCTL 0x36   ///< Memory Access Control
+#define GC9A01A_VSCRSADD 0x37 ///< Vertical Scrolling Start Address
+#define GC9A01A_PIXFMT 0x3A   ///< COLMOD: Pixel Format Set
```
Comment (jfischer-no) [inline drivers/display/display_gc9a01a.h:42]: Usually when writing drivers, especially the first one, it is a good idea to use an existing one as a template. There would be copyright notices from the original or forgotten comments to change, but here `///< COLMOD: Pixel Format Set` looks a bit confusing. Searching for this pattern we find [Adafruit_GC9A01A.h](https://github.com/adafruit/Adafruit_GC9A01A/commit/763dc26bfe95276044fdb1b8c12a08cb0df9e1d9#diff-0e1829dee3da6a0668fb2a86074cd36e018f139b89dd906dfaa1445d87ba20e5)
adafruit_gc9a01a
This is not acceptable.

File: drivers/display/display_gc9a01a.c [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+/**
+ * @brief Transmit values to the display driver
+ *
+ * @param dev GC9A01A device instance.
+ * @param cmd command associated with the register.
+ * @param tx_data Data to be sent.
+ * @param tx_len Length of buffer to be sent.
+ * @return 0 on success, errno otherwise.
+ */
+int gc9a01a_transmit(const struct device *dev, uint8_t cmd, const void *tx_data,
```
Comment (jfischer-no) [inline drivers/display/display_gc9a01a.c]: `static int gc9a01a_transmit(const struct device *dev, uint8_t cmd, const void *tx_data,`
There should not be any doxygen comments in this file.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

**Comment by jfischer-no** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/62934#discussion_r1334416979
Closing as not acceptable.

File: drivers/display/display_gc9a01a.h (hunk 10 above)
Comment (jfischer-no) [reply drivers/display/display_gc9a01a.h:42]: https://docs.zephyrproject.org/latest/contribute/guidelines.html#licensing It is not worth not having an Apache license for it as you can type these defines in relatively short time from the datasheet.

**Comment by joelguittet** [comment]
@amrithvenkat @jfischer-no why stopping working on this driver ?

**Comment by ljunquera** [comment]
There seems to be a lot of interest in this driver and it seems like it's close. Is there anybody who can help finish it off from the community?

**Comment by joelguittet** [comment]
@ljunquera for sure, I m actually trying to have zephyr working on the esp32s3-touch-lch-1.28 from waveshare. 
I have difficulties with building lvgl and driver together, looks like it s more related to the integration with the board or bootloader, not sure....
The board has also accelerator/gyroscope that would be nice to integrate in zephyr too.
Will be happy to collaborate with other people on this, please ping if you are interested or currently working on this :-)

**Comment by ljunquera** [comment]
@joelguittet I would love to help but am a real novice. I am fighting though some of the cmake stuff right now. I know @jakkra has it working in ZSWatch. However I can help I will.

**Comment by amrithvenkat** [comment]
@joelguittet @ljunquera I would like for the driver to be included in the code base. Even @jakkra has the implementation which is also similar to my PR. I face licensing issues which I'll be glad If anyone helps me out to resolve. Thanks

**Comment by joelguittet** [comment]
@jakkra implementation need few improvement as far as I have seen. For example it is limiting one display, instead of foreach each DTS entry.
By the way I agree this is a good starting point, I m actually using his driver (or at least I try!)
So far I m able to build simple app for esp32s3 but as soon as I'm adding some specific lvgl stuff (and particularly the function lv_scr_act) the binary becomes really bigger (60kB) and the application doesn't boot at all. The device remains stuck at the end of the bootloader.
@ljunquera @amrithvenkat do you have the esp32s3 touch LCD hardware from waveshare or something else ?

**Comment by ljunquera** [comment]
I'm trying to build it for the nRF5340 DK using the lvgl sample as the base. I have the [Waveshare 1.28inch touch LCD](https://www.waveshare.com/wiki/1.28inch_Touch_LCD) and the nRF Connect for VS Add-In. My current problems are basic (CMakeLists and overlay node)...again I'm new to this.

**Comment by amrithvenkat** [comment]
@ljunquera I used the same sample from waveshare as the reference. And I was able to use multiple displays with my implementation.
@joelguittet implementation I did was on the Rpi pico rp2040. Unfortunately I don't have sample if it works with esp or nrf. Even ZSWatch implementation is on rp2040.

**Comment by ljunquera** [comment]
Thanks, @amrithvenkat. I will try to get it working with nRF.

**Comment by joelguittet** [comment]
Great example @amrithvenkat 
I have not rpi pico but will try as well, on the esp32s3.
I see you have not lvgl, looks a good step first to test the display.

About licensing issues I think the main solution is that the driver should be in "zephyr style". It s true we can't just copy from another source, but of course the goal is not to reinvent everything, it's possible to get inspiration from other examples from Arduino, esp IDF etc
My approach will be more or less to copy an existing and similar driver from zephyr, and modify where it is required.

**Comment by ljunquera** [comment]
@joelguittet, what's a good driver to start with which is already in Zephyr?

**Comment by joelguittet** [comment]
@ljunquera the one with spi bus, command/data and reset gpio will be a good starting point. St7789 for example?
This has also the advantage to not create new convention, for example cd and rst gpio should be called with the same name, in DTS and in the source too.

**Comment by MrMarteng** [comment]
Hi guys,
I also started working on this driver as part of support for m5stack m5dial. I didn't have access to the initial branch therefore I started reworking requests on a separate branch: 
https://github.com/MrMarteng/zephyr/tree/gc9a91a_display

**Comment by ljunquera** [comment]
Thanks, @MrMarteng. Great work. Are we going to run into the same copyright issues which caused a problem in the first pass?

**Comment by MrMarteng** [comment]
> Thanks, @MrMarteng. Great work. Are we going to run into the same copyright issues which caused a problem in the first pass? Should we start with what @joelguittet recommended and start with a new build based on the existing st7789 and a copyright consistent with Zephyr Project.

Yes I saw the issue. I recreated the according files from scratch. Give me a few hours, I will push the latest version to the mentioned branch.

**Comment by joelguittet** [comment]
@MrMarteng awesome! Will check it soon

**Comment by ljunquera** [comment]
It would be great to have a sample to go with it.

**Comment by amrithvenkat** [comment]
Work on this driver continued on PR - #67012

**Pull request closed without merging**